tools/infrasys/src/main.rs

```diff
@@ -0,0 +1,767 @@
+
+    // Push the new policy
+    s3_client
+        .put_bucket_policy(PutBucketPolicyRequest {
+            bucket: bucket_name.clone(),
+            policy: String::from(current_bp),
+            ..Default::default()
+        })
+        .await
+        .context(error::PutPolicy { bucket_name })?;
+
+    Ok(())
+}
+
+/// Uploads root.json to S3 Bucket
+/// Input: Region and name of bucket, which folder root.json should be pushed to, and vpcid
+/// Note that the prefix parameter must have the format "/..." and the bucket name "..."
+/// Behavior: This function will either create the folder that the bucket policy was scoped to, or if the folder exists, simply add to it
+/// Output: Doesn't need to save any metadata from this action
+async fn upload_file(
+    region: &String,
+    bucket_name: &String,
+    prefix: &String,
+    file_path: &PathBuf,
+) -> Result<()> {
+    let s3_client =
+        S3Client::new(Region::from_str(region).context(error::ParseRegion { what: region })?);
+
+    // File --> Bytes
+    let mut file = File::open(file_path).context(error::FileOpen { path: file_path })?;
+    let mut buffer = Vec::new();
+    file.read_to_end(&mut buffer)
+        .context(error::FileRead { path: file_path })?;
+
+    s3_client
+        .put_object(PutObjectRequest {
+            bucket: format!("{}{}", bucket_name, prefix),
+            key: "root.json".to_string(),
+            body: Some(StreamingBody::from(buffer)),
+            ..Default::default()
+        })
+        .await
+        .context(error::PutObject { bucket_name })?;
+
+    Ok(())
+}
+
+//  =^..^=   =^..^=   =^..^= KEY GENERATION  =^..^=   =^..^=   =^..^=
+
+/// Returns the available keys field of a SigningKeyConfig enum. Available keys represents either all the root keys or publication keys
+fn get_available_keys(signing_key_config: &SigningKeyConfig) -> Result<HashMap<String, String>> {
+    match signing_key_config {
+        SigningKeyConfig::file { .. } => Ok(HashMap::new()),
+        // If available keys is empty here, it means keys were not created properly
+        SigningKeyConfig::kms { config, .. } => Ok(config
+            .available_keys
+            .as_ref()
+            .context(error::KeyCreation)?
+            .clone()),
+        SigningKeyConfig::ssm { .. } => Ok(HashMap::new()),
+    }
+}
+
+/// Creates keys using data stored in SigningKeyConfig enum
+/// Output: This method will has an in-place output meaning it edits key-id
+/// If key-id is populated, it will not change. Otherwise, it will be populated with a key-id of an available key
+async fn create_keys(signing_key_config: &mut SigningKeyConfig) -> Result<()> {
+    match signing_key_config {
+        SigningKeyConfig::file { path } => (),
+        SigningKeyConfig::kms { key_id, config, .. } => {
+            // QUESTION: keyid must remain at top level (we could encapsulate in config, but it would alter keyid reading for pubsys (which we alredy are technically doing by changing it to Option))
```

Comment:
We should probably just keep the `key_id` field outside ot `KMSConfig` if that's what you're referring to.